compiler/rustc_parse/src/parser/item.rs

```diff
@@ -1724,6 +1724,20 @@ impl<'a> Parser<'a> {
                 err.help("unlike in C++, Java, and C#, functions are declared in `impl` blocks");
                 err.help("see https://doc.rust-lang.org/book/ch05-03-method-syntax.html for more information");
                 err
+            } else if self.check_keyword(kw::Struct) {
+                let kw_token = self.token.clone();
+                let item = self.parse_item(ForceCollect::No)?;
```

Comment:
The implementation would eventually look something like the following:

```rust
} else if self.eat_keyword(kw::Struct) {
    match self.parse_item_struct() {
        Ok((ident, _)) => {
            let mut err = self.struct_span_err(
                lo.with_hi(ident.span.hi()),
                &format!("structs are not allowed in {adt_ty} definitions"),
            );
            err.help("consider creating a new `struct` definition instead of nesting");
            err
        }
        Err(err) => {
            err.cancel();
            self.restore_snapshot(snapshot);
            self.expected_ident_found()
        }
    }
}
```